### PR TITLE
AST-based operations

### PR DESCRIPTION
**Pull request by klavs**
Allow creating operations out of AST.
This is a non-breaking transitional to enable AST-based links, local state resolvers.

This change add as slight performance reduction because documents must be printed to string before sending them over the network.
The remedy for that is building AST at build-time. Currently that is possible using `package:gql_code_gen`.

### CODE REVIEW

**Comment by klavs** [comment]
@HofmannZ, here's the non-breaking change I was referring to in https://github.com/zino-app/graphql-flutter/pull/414

File: packages/graphql/lib/src/core/query_options.dart [hunk 1]
```diff
@@ -67,12 +69,17 @@ class Policies {
 /// Base options.
 class BaseOptions extends RawOperationData {
   BaseOptions({
-    @required String document,
+    @Deprecated('Use documentNode instead') String document,
```
Comment (HofmannZ) [inline packages/graphql/lib/src/core/query_options.dart]: Good to first deprecated support for `document`. Maybe rephrase the message to:

```
'The "document" option has been deprecated, use "documentNode" instead'
```

File: packages/graphql/lib/src/core/query_options.dart [hunk 2]
```diff
@@ -194,12 +207,25 @@ typedef dynamic UpdateQuery(
...
+    this.documentNode,
     this.variables = const <String, dynamic>{},
     @required this.updateQuery,
-  }) : assert(updateQuery != null);
+  })  : assert((document != null && documentNode == null) ||
+            (document == null && documentNode != null)),
+        assert(updateQuery != null);
+
+  DocumentNode documentNode;
```
Comment (HofmannZ) [inline packages/graphql/lib/src/core/query_options.dart]: Why not use getters and setters?

File: packages/graphql/lib/src/core/raw_operation_data.dart [hunk 3]
```diff
@@ -31,7 +45,7 @@ class RawOperationData {
   String get operationName {
     // XXX there is a bug in the `graphql_parser` package, where this result might be
     // null event though the operation name is present in the document
-    _operationName ??= getOperationName(document);
+    _operationName ??= getLastOperationName(documentNode);
```
Comment (HofmannZ) [inline packages/graphql/lib/src/core/raw_operation_data.dart]: Can we test if this bug is fixed?

File: packages/graphql/lib/src/core/query_options.dart (hunk 1 above)
Comment (klavs) [reply packages/graphql/lib/src/core/query_options.dart]: OK, added a more verbose deprecation message

File: packages/graphql/lib/src/core/query_options.dart (hunk 2 above)
Comment (klavs) [reply packages/graphql/lib/src/core/query_options.dart]: Added the missing initialisation to the constructor

File: packages/graphql/lib/src/core/raw_operation_data.dart (hunk 3 above)
Comment (klavs) [reply packages/graphql/lib/src/core/raw_operation_data.dart]: Removing the comment. It's not too helpful in describing the exact conditions of the bug.

If any bug in parser appears, we'll file a bug to `package:gql` and we'll fix it there.

**Comment by klavs** [comment]
@HofmannZ I addressed the feedback and signed the commits.

I think the CI fails incorrectly – we should be able to deprecate our own API and use it while it is deprecated.

cc: @mainawycliffe

**Comment by klavs** [comment]
added info about AST documents to the readme.

Note that I'm assuming the version `2.2.0`

**Comment by lookfirst** [comment]
This is awesome!
